### PR TITLE
GET /api/v1/appointments/search

### PR DESCRIPTION
**Pull request by Alaalser**
Relates #13 
## GET /api/v1/appointments/search
- Create a query that can search for name or phone or both
- Create its query and controller
- Create its router
- Create query tests and route tests for both success and failed

### CODE REVIEW

File: server/database/queries/appointments/getAppointmentsByPatientName.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+const connection = require('../../config/connection');
+
+const getAppointmentsByPatientName = ({ firstName, lastName, phone }) => {
+  const sql = {
+    text:
+      'SELECT appointments.id,appointments.appointment_date,appointments.appointment_time,patients.firstname,patients.lastname FROM appointments INNER JOIN patients ON patients.id = appointments.patient_id where appointments.is_done =false AND patients.firstname = $1 AND patients.lastname = $2 OR (patients.phone= $3 AND appointments.is_done =false) ;',
```
Comment (hassanelnajjar) [inline server/database/queries/appointments/getAppointmentsByPatientName.js:6]: Please select the patient phone also, Assume the patient time come and he didn't arrive and the doctor wants to call him  then He want the phone number :)

File: server/controllers/appointments/appointmentsSearch.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+      abortEarly: false,
+    });
+
+    if ((firstName && lastName) || phone) {
+      const { rows: data } = await getAppointmentsByPatientName({
+        firstName,
+        lastName,
+        phone,
+      });
+      res.json({
```
Comment (hassanelnajjar) [inline server/controllers/appointments/appointmentsSearch.js:21]: Alaa,when I am using your routes like this :-
http://localhost:5000/api/v1/appointments/search?firstName=Alexie&lastName=Jenkins
I got 3 items but I expected one .

```json
{
StatusCode: 200,
data: [
{
id: 1,
appointment_date: "2021-12-02T00:00:00.000Z",
appointment_time: "22:30:00",
firstname: "Alexie",
lastname: "Jenkins"
},
{
id: 4,
appointment_date: "2021-12-02T00:00:00.000Z",
appointment_time: "22:30:00",
firstname: "Alexie",
lastname: "Jenkins"
},
{
id: 8,
appointment_date: "2020-12-02T00:00:00.000Z",
appointment_time: "20:30:00",
firstname: "Alexie",
lastname: "Jenkins"
}
]
}
```

File: server/controllers/appointments/appointmentsSearch.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  try {
+    const {
+      firstName,
+      lastName,
+      phone,
+    } = await appointmentsSearchSchema.validate(req.query, {
+      abortEarly: false,
+    });
+
+    if ((firstName && lastName) || phone) {
```
Comment (hassanelnajjar) [inline server/controllers/appointments/appointmentsSearch.js:15]: I think you don't need to link the first name with last name by && operator, it's good to let the route work by supplying any of these three values.

**Review by hassanelnajjar — CHANGES_REQUESTED** [review]

File: package.json [hunk 4]
```diff
@@ -58,4 +58,4 @@
     "eslint-plugin-prettier": "^3.3.1",
     "prettier": "2.2.1"
   }
-}
+}
```
Comment (Mu7ammadAbed) [inline package.json:61]: new line here

File: server/routes/appointments/index.js [hunk 5]
```diff
@@ -1,6 +1,9 @@
 const router = require('express').Router();
+const { appointmentsSearch } = require('../../controllers');
 const { getAppointmentsByDate } = require('../../controllers');
 
+router.get('/search', appointmentsSearch);
+router.get('/:date');
 router.get('/search');
```
Comment (Mu7ammadAbed) [inline server/routes/appointments/index.js:7]: Don't repeat routes and route `/:date` is used for what?

File: server/utils/validation/index.js [hunk 6]
```diff
@@ -1,3 +1,8 @@
+const appointmentsSearchSchema = require('./appointmentsSearchSchema');
+
+module.exports = {
+  appointmentsSearchSchema,
+};
```
Comment (Mu7ammadAbed) [inline server/utils/validation/index.js:5]: change the way you are exporting modules with it to the way your teammates using it.

File: server/routes/appointments/index.js (hunk 5 above)
Comment (Alaalser) [reply server/routes/appointments/index.js:7]: It's from file structure !!!

File: server/database/queries/getAppointmentsByPatientName/index.js [hunk 7]
```diff
@@ -0,0 +1,12 @@
+const connection = require('../../config/connection');
+
+const getAppointmentsByPatientName = ({ firstName, lastName, phone }) => {
+  const sql = {
+    text:
+      'SELECT appointments.id,appointments.appointment_date,appointments.appointment_time,patients.firstname,patients.lastname,patients.phone FROM appointments INNER JOIN patients ON patients.id = appointments.patient_id WHERE appointments.is_done =false AND (patients.firstname ILIKE $1 OR patients.lastname ILIKE $2 OR patients.phone= $3);',
```
Comment (Mu7ammadAbed) [inline server/database/queries/getAppointmentsByPatientName/index.js:6]:
```suggestion
      'SELECT appointments.id AS appointment_id, patients.id AS patient_id,appointments.appointment_date,appointments.appointment_time, patients.firstname,patients.lastname,patients.phone FROM appointments INNER JOIN patients ON patients.id = appointments.patient_id WHERE appointments.is_done =false AND (patients.firstname ILIKE $1 OR patients.lastname ILIKE $2 OR patients.phone= $3);',
```

**Review by Mu7ammadAbed — APPROVED** [review]

**Review by Mu7ammadAbed — APPROVED** [review]

**Pull request merged**